File: C#/XamarinStore.Droid/Fragments/ShippingDetailsFragment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Util;
using Android.Views;
using Android.Widget;
using Android.Views.InputMethods;
using System.Threading.Tasks;

namespace XamarinStore
{
	public class ShippingDetailsFragment : Fragment
	{
		User user;

		public ShippingDetailsFragment() : this(new User()) {}

		public ShippingDetailsFragment(User user)
		{
			this.user = user;
		}

		public override void OnCreate (Bundle savedInstanceState)
		{
			base.OnCreate (savedInstanceState);
			RetainInstance = true;
		}

		AutoCompleteTextView state;
		AutoCompleteTextView country;
		public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
		{
			var shippingDetailsView = inflater.Inflate(Resource.Layout.ShippingDetails, container, false);

			var placeOrder = shippingDetailsView.FindViewById<Button> (Resource.Id.PlaceOrder);
			var phone = shippingDetailsView.FindViewById<EditText> (Resource.Id.Phone);
			phone.Text = user.Phone;

			var firstName = shippingDetailsView.FindViewById<EditText> (Resource.Id.FirstName);
			firstName.Text = user.FirstName;

			var lastName = shippingDetailsView.FindViewById<EditText> (Resource.Id.LastName);
			lastName.Text = user.LastName;

			var streetAddress1 = shippingDetailsView.FindViewById<EditText> (Resource.Id.StreetAddress1);
			streetAddress1.Text = user.Address;

			var streetAddress2 = shippingDetailsView.FindViewById<EditText> (Resource.Id.StreetAddress2);
			streetAddress2.Text = user.Address2;

			var city = shippingDetailsView.FindViewById<EditText> (Resource.Id.City);
			city.Text = user.City;

			state = shippingDetailsView.FindViewById<AutoCompleteTextView> (Resource.Id.State);
			state.Text = user.State;

			var postalCode = shippingDetailsView.FindViewById<EditText> (Resource.Id.PostalCode);
			postalCode.Text = user.ZipCode;

			country = shippingDetailsView.FindViewById<AutoCompleteTextView> (Resource.Id.Country);
			user.Country = string.IsNullOrEmpty (user.Country) ? "United States" : user.Country;
			country.Text = user.Country;
			country.ItemSelected += (object sender, AdapterView.ItemSelectedEventArgs e) => {
				LoadStates();
			};

			placeOrder.Click += async (sender, e) => {
				var entries = new EditText[] {
					phone, streetAddress1, streetAddress2, city, state, postalCode, country
				};
				foreach (var entry in entries)
					entry.Enabled = false;
				user.FirstName = firstName.Text;
				user.LastName = lastName.Text;
				user.Phone = phone.Text;
				user.Address = streetAddress1.Text;
				user.Address2 = streetAddress2.Text;
				user.City = city.Text;
				user.State = state.Text;
				user.ZipCode = postalCode.Text;
				user.Country = await WebService.Shared.GetCountryCode(country.Text);
				await ProcessOrder();
				foreach (var entry in entries)
					entry.Enabled = true;
			};
			LoadCountries ();
			LoadStates ();
			return 	shippingDetailsView;
		}

		async void LoadCountries()
		{
			var countries = await WebService.Shared.GetCountries ();
			country.Adapter = new ArrayAdapter(this.Activity, Android.Resource.Layout.SimpleDropDownItem1Line, countries.Select(x=> x.Name).ToList());
		}

		async void LoadStates()
		{
			var states = await WebService.Shared.GetStates (country.Text);
			state.Adapter = new ArrayAdapter(this.Activity, Android.Resource.Layout.SimpleDropDownItem1Line, states);
		}

		async Task ProcessOrder ()
		{	
			var isValid = await user.IsInformationValid ();
			if (!isValid.Item1) {
				Toast.MakeText (Activity, isValid.Item2, ToastLength.Long).Show ();
				return;
			}

			var progressDialog = ProgressDialog.Show(this.Activity, "Please wait...", "Placing Order", true);
			var result = await WebService.Shared.PlaceOrder (user);
			progressDialog.Hide ();
			progressDialog.Dismiss ();
			string message = result.Success ? "Your order has been placed!" : "Error: " + result.Message;
			Toast.MakeText (Activity, message, ToastLength.Long).Show ();
			if (!result.Success)
				return;
			var op = OrderPlaced;
			if (op != null)
				op ();
		}

		public Action OrderPlaced {get;set;}
	}

}

File: F#/XamarinStore.Droid/Fragments/ShippingDetailsFragment.fs
namespace XamarinStore

open System
open System.Linq

open Android.App
open Android.OS
open Android.Views
open Android.Widget
open System.Threading.Tasks
open Helpers

type ShippingDetailsFragment(userParam: User.User) as this =
    inherit Fragment()
    let mutable user = userParam
    let mutable state:AutoCompleteTextView = null
    let mutable country:AutoCompleteTextView = null

    let LoadCountries() = async {
        let! countries = WebService.Shared.GetCountries ()
        country.Adapter <- new ArrayAdapter(this.Activity, Android.Resource.Layout.SimpleDropDownItem1Line, countries.Select(fun (x:WebService.Country) -> x.Name).ToList()) :> IListAdapter }

    let LoadStates() =
        let states = WebService.Shared.GetStates(country.Text).ToArray()
        state.Adapter <- (new ArrayAdapter(this.Activity, Android.Resource.Layout.SimpleDropDownItem1Line, states) :> IListAdapter)

    let ProcessOrder () = async {   
        let! result = WebService.Shared.ValidateUser user
        match result with
        | Failure msg -> Toast.MakeText(this.Activity, msg, ToastLength.Long).Show()
        | Success _ ->
            let progressDialog = ProgressDialog.Show(this.Activity, "Please wait...", "Placing Order", true)
            let! result = WebService.Shared.PlaceOrder user
            progressDialog.Hide ()
            progressDialog.Dismiss ()
            let message = if result.Success 
                          then "Your order has been placed!"
                          else "Error: " + result.Message
            Toast.MakeText(this.Activity, message, ToastLength.Long).Show()
            if result.Success then
                this.OrderPlaced() }

    override this.OnCreate savedInstanceState =
        base.OnCreate savedInstanceState
        this.RetainInstance <- true

    override this.OnCreateView(inflater, container, savedInstanceState) =
        let shippingDetailsView = inflater.Inflate(Resource_Layout.ShippingDetails, container, false)

        let placeOrder = shippingDetailsView.FindViewById<Button> (Resource_Id.PlaceOrder)
        let phone = shippingDetailsView.FindViewById<EditText> (Resource_Id.Phone)
        phone.Text <- user.Phone

        let firstName = shippingDetailsView.FindViewById<EditText> (Resource_Id.FirstName)
        firstName.Text <- user.FirstName

        let lastName = shippingDetailsView.FindViewById<EditText> (Resource_Id.LastName)
        lastName.Text <- user.LastName

        let streetAddress1 = shippingDetailsView.FindViewById<EditText> (Resource_Id.StreetAddress1)
        streetAddress1.Text <- user.Address

        let streetAddress2 = shippingDetailsView.FindViewById<EditText> (Resource_Id.StreetAddress2)
        streetAddress2.Text <- user.Address2
       
        let city = shippingDetailsView.FindViewById<EditText> (Resource_Id.City)
        city.Text <- user.City

        state <- shippingDetailsView.FindViewById<AutoCompleteTextView> (Resource_Id.State)
        state.Text <- user.State

        let postalCode = shippingDetailsView.FindViewById<EditText> (Resource_Id.PostalCode)
        postalCode.Text <- user.ZipCode

        country <- shippingDetailsView.FindViewById<AutoCompleteTextView> (Resource_Id.Country)
        user <- { user with Country = if user.Country = null then "United States" else user.Country }
        country.Text <- user.Country
        country.ItemSelected.Add(fun _-> LoadStates())

        placeOrder.Click.Add(fun _ -> async {
                                        try
                                            let entries = [| phone
                                                             streetAddress1
                                                             streetAddress2
                                                             city
                                                             state:> EditText
                                                             postalCode
                                                             country :> EditText |]
                                            for entry in entries do
                                                entry.Enabled <- false
                                            
                                            let! countryCode = WebService.Shared.GetCountryCodeFromName(country.Text)

                                            user <- { user with 
                                                        FirstName = firstName.Text
                                                        LastName = lastName.Text
                                                        Phone = phone.Text
                                                        Address = streetAddress1.Text
                                                        Address2 = streetAddress2.Text
                                                        City = city.Text
                                                        State = state.Text
                                                        ZipCode = postalCode.Text
                                                        Country = countryCode }
                                            do! ProcessOrder()
                                            for entry in entries do
                                                entry.Enabled <- true
                                        with e -> ()
                                            } |> Async.StartImmediate)
        LoadCountries () |> Async.StartImmediate
        LoadStates ()
        shippingDetailsView

    member val OrderPlaced = fun ()->() with get,set
